resilience4j-timelimiter/src/main/java/io/github/resilience4j/timelimiter/internal/TimeLimiterImpl.java

```diff
@@ -6,27 +6,37 @@
 import io.github.resilience4j.timelimiter.event.TimeLimiterOnErrorEvent;
 import io.github.resilience4j.timelimiter.event.TimeLimiterOnSuccessEvent;
 import io.github.resilience4j.timelimiter.event.TimeLimiterOnTimeoutEvent;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.LongAdder;
 import java.util.function.Supplier;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 public class TimeLimiterImpl implements TimeLimiter {
 
     private static final Logger LOG = LoggerFactory.getLogger(TimeLimiterImpl.class);
 
     private String name;
     private final TimeLimiterConfig timeLimiterConfig;
+    private final LongAdder successes;
+    private final LongAdder errors;
+    private final LongAdder timeouts;
+    private final TimeLimiterMetrics metrics;
     private final TimeLimiterEventProcessor eventProcessor;
 
     public TimeLimiterImpl(String name, TimeLimiterConfig timeLimiterConfig) {
         this.name = name;
         this.timeLimiterConfig = timeLimiterConfig;
+        this.successes = new LongAdder();
```

Comment:
I've a mixed feeling if we should really measure successful, failed and timeout calls here.
It means a TimeLimiter measures calls even if the user is not interested in the metrics. Other components like RateLimiter, CircuitBreaker or Bulkhead can provide some internal metrics, because they need some them anyway to function properly. And other some other metrics of the CircuitBreaker or RateLimiter are measured by consiming the events. Please see [TaggedCircuitBreakerMetrics](https://github.com/resilience4j/resilience4j/blob/master/resilience4j-micrometer/src/main/java/io/github/resilience4j/micrometer/tagged/TaggedCircuitBreakerMetrics.java#L130-L134).

I think I would prefer that a TimeLimiter itself is lightweight and does not measure metrics itself. The `TimeLimiterMetrics` class can register event consumers and increases counters.

What is your opinion?
@Romeh @hexmind @storozhukBM

Reply:
You are right, only `cache` and `retry` components record and store metrics. Maybe we can provide metrics on demand only if the user is not interested in the metrics? Something like below:
```
TimeLimiterImpl
// ...
@Override
    public Metrics createMetrics() {
        return new Metrics() {
            private final LongAdder successes = new LongAdder();
            private final LongAdder errors = new LongAdder();
            private final LongAdder timeouts = new LongAdder();

            {
                getEventPublisher().onSuccess(e -> successes.increment());
                getEventPublisher().onError(e -> errors.increment());
                getEventPublisher().onTimeout(e -> timeouts.increment());
            }

            @Override
            public long getNumberOfSuccessfulCalls() {
                return successes.longValue();
            }

            @Override
            public long getNumberOfFailedCalls() {
                return errors.longValue();
            }

            @Override
            public long getNumberOfTimedOutCalls() {
                return timeouts.longValue();
            }
        };
    }
```

Reply:
I would prefer to implement it similar to the other components and register event consumers in the Dropwizard Metrics / MicroMeter / Prometheus implementations.

Reply:
I will keep only an interface in time limiter and add implementations in proper modules later.

Reply:
You can delete the interface. The metrics module is independent.